src/Functions/dateTimeToSnowflakeID.cpp

```diff
@@ -0,0 +1,181 @@
+
+    static FunctionPtr create(ContextPtr context) { return std::make_shared<FunctionDateTime64ToSnowflakeID>(context); }
+    explicit FunctionDateTime64ToSnowflakeID(ContextPtr context)
+        : uniform_snowflake_conversion_functions(context->getSettingsRef().uniform_snowflake_conversion_functions)
+    {}
+
+    String getName() const override { return name; }
+    size_t getNumberOfArguments() const override { return 0; }
+    bool isVariadic() const override { return true; }
+    bool useDefaultImplementationForConstants() const override { return true; }
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        FunctionArgumentDescriptors args{
+            {"value", static_cast<FunctionArgumentDescriptor::TypeValidator>(&isDateTime64), nullptr, "DateTime64"}
+        };
+        FunctionArgumentDescriptors optional_args{
+            {"epoch", static_cast<FunctionArgumentDescriptor::TypeValidator>(&isNativeUInt), isColumnConst, "UInt*"}
+        };
+        validateFunctionArgumentTypes(*this, arguments, args, optional_args);
+
+        return std::make_shared<DataTypeUInt64>();
+    }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+    {
+        if (!uniform_snowflake_conversion_functions)
+            throw Exception(ErrorCodes::UNKNOWN_FUNCTION, "To use function {}, setting 'uniform_snowflake_conversion_functions' must be enabled", getName());
+
+        const auto & col_src = *arguments[0].column;
+        const auto & src_data = typeid_cast<const ColumnDateTime64 &>(col_src).getData();
+
+        size_t epoch = 0;
+        if (arguments.size() == 2 && input_rows_count != 0)
+        {
+            const auto & col_epoch = *arguments[1].column;
+            epoch = col_epoch.getUInt(0);
+        }
+
+        auto col_res = ColumnUInt64::create(input_rows_count);
+        auto & res_data = col_res->getData();
+
+        /// timestamps in snowflake-ids are millisecond-based, convert input to milliseconds
+        UInt32 src_scale = getDecimalScale(*arguments[0].type);
+        Int64 multiplier_msec = DecimalUtils::scaleMultiplier<DateTime64>(3);
+        Int64 multiplier_src = DecimalUtils::scaleMultiplier<DateTime64>(src_scale);
+        auto factor = multiplier_msec / static_cast<double>(multiplier_src);
+
+        for (size_t i = 0; i < input_rows_count; ++i)
+            res_data[i] = static_cast<UInt64>(src_data[i] * factor - epoch) << time_shift;
+
+        return col_res;
+    }
+};
+
+REGISTER_FUNCTION(DateTimeToSnowflakeID)
+{
+    {
+        FunctionDocumentation::Description description = R"(Converts a [DateTime](../data-types/datetime.md) value to the first [Snowflake ID](https://en.wikipedia.org/wiki/Snowflake_ID) at the giving time.)";
```

Comment:
Is the markdown style make sense here? Isn't pure text better? I think these docs are not part of any website/pretty printed things, only returned from `system.functions`, thus for me the MD style formatting doesn't make sense. However feel free to keep it, it is not a big deal, I have no strong opinion about them.

Reply:
Your point makes sense and it points to a generally open question for which no good solution exists yet. Until we have tooling which generates markdown docs from in-source docs (<-- which will ideally require some link format for embedded docs that can be displayed in non-markdown format in system.functions and converted into markdown format), we can keep in-source and markdown docs at least consistent.